### ISSUE TITLE
[ROCm/MI325X] DeepSeek-V4-Flash: Triton fp8_mqa_logits kernel requires 96KB shared memory, MI325X limit is 64KB

### ISSUE THREAD

**Issue by angelopaolosg** [vllm-project/vllm#41963]
## Environment

- **Hardware**: AMD MI325X × 8 (255.98 GiB HBM each)
- **vLLM version**: 0.20.2rc1.dev90+g7a576e2c7 (rocm/vllm-dev:nightly, 2026-05-06)
- **Model**: deepseek-ai/DeepSeek-V4-Flash
- **Config**: TP=8, EP=8, `--kv-cache-dtype fp8`, `--enforce-eager`

## Bug

At inference time, the first request fails with a Triton shared memory OOM:

```
triton.runtime.errors.OutOfResources: out of resource: shared memory,
Required: 98304, Hardware limit: 65536.
Reducing block sizes or `num_stages` may help.
```

Call stack:
```
rocm_aiter_mla_sparse.py:538 → rocm_fp8_mqa_logits
aiter/ops/triton/attention/fp8_mqa_logits.py:53 → _fp8_mqa_logits_kernel
```

## Root Cause

The Triton kernel `_fp8_mqa_logits_kernel` in `aiter/ops/triton/attention/fp8_mqa_logits.py` is configured with block sizes that require **96 KB** of LDS (Local Data Share / shared memory). The AMD MI325X GPU supports a maximum of **64 KB** per workgroup.

The PR #40871 was developed and tested exclusively on **MI355X**, which appears to support the larger shared memory configuration required by this kernel.

## Impact

DeepSeek-V4-Flash is completely non-functional for inference on MI325X. The model loads correctly but crashes on the first forward pass.

## Suggested Fix

Add a hardware-capability check before launching the kernel, and reduce `BLOCK_SIZE` or `num_stages` for GPUs with ≤64 KB shared memory. The fix should be in `aiter` (the AITER library used by vLLM ROCm).

## Related

- PR #40871 "Add AMD support for DeepSeek V4" (merged 2026-05-05) — tested on MI355X only
- Issue #41961: `mul_cuda` not implemented for `Float8_e8m0fnu` (MI325X)
- Issue #41962: OOM in `rocm_dequantize_blocked_k_cache` + CUDAGraph incompatibility

**Comment by msunda17** [comment]
Hi,

I noticed the shared memory pressure on the fp8_mqa_logits Triton kernel. One common source of subtle correctness issues in new Flash-style kernels is silent errors in block mapping/memory routing that unit tests and shape checks miss.

I built a free open-source CLI called 'ImpactArbiter' that runs deterministic gradient-based verification on KV-cache / block-mapping code (traps the function in an autograd graph and compares against a SymPy Oracle). It catches exactly the kind of mathematical hallucinations that only show up under real load.

Happy to run `impactarbiter audit` on any candidate kernel you have and share the divergence maps / auto-heal results if useful. Repo: https://github.com/msunda17/impactarbiter-cli

Let me know!

**Comment by vllmellm** [comment]
Hi, I've tested deepseek-v4-flash on mi300x in the latest upstream vllm image `vllm/vllm-openai-rocm:v0.27.1`. 

This command works:
```
export VLLM_ROCM_USE_AITER=1

vllm serve deepseek-ai/DeepSeek-V4-Flash-0731 \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --tensor-parallel-size 8 \
  --distributed-executor-backend mp \
  --gpu-memory-utilization 0.9 \
  --max-num-seqs 512 \
  --max-num-batched-tokens 8192 \
  --compilation-config '{"mode": 3, "cudagraph_mode": "FULL_DECODE_ONLY"}' \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4 \
  --reasoning-config '{"reasoning_parser":"deepseek_v4","reasoning_start_str":"","reasoning_end_str":""}'
```

However, the `--enable-expert-parallel` seems to be producing erratic result, we will take a look into it.

**Comment by vllmellm** [comment]
Using `vllm/vllm-openai-rocm:nightly-d626108b1841888ec90aced33367149a6bbc7e4b` EP is working properly too.
